dali/python/nvidia/dali/pipeline.py

```diff
@@ -1884,6 +1885,98 @@ def create_pipeline(*args, **kwargs):
     return actual_decorator(fn) if fn else actual_decorator
 
 
+# Callable preserving a signature
+_F = TypeVar("_F", bound=Callable[..., Any])
+
+
+def do_not_convert(func: _F = None) -> _F:
+    """Decorator that suppresses the conversion of a function by AutoGraph.
+
+    In conditional mode, DALI uses a fork of
+    `TensorFlow's AutoGraph <https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/autograph/g3doc/reference/index.md>`_
+    to transform the code, enabling us to rewrite and detect the ``if`` statements, so they can be
+    used in processing the DALI pipeline.
+
+    When used with :meth:`external source <nvidia.dali.fn.external_source>` in parallel mode
+    (``parallel=True``), this may interfere with the serialization of the provided ``source``
+    parameter. To prevent this, functions that are used to create the ``source`` parameter,
+    should be decorated with :meth:`@do_not_convert <nvidia.dali.pipeline.do_not_convert>`.
```

Comment:
All that is true - but on the other hand, this decorator is in `pipeline` module, not `external_source`, so the focus on a particular flavor of external source seems exaggerated. I know that this was a precipitating factor for the whole effort, but it has wider implications and this documentation seems a bit awkward and misplaced when read without context. It also leaves out other potentially problematic areas, like PythonFunction (perf. hit) or numba (do we know if this works?).
```suggestion
    In conditional mode, DALI uses a fork of
    `TensorFlow's AutoGraph <https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/autograph/g3doc/reference/index.md>`_
    to transform the code, enabling us to rewrite and detect the ``if`` statements, so they can be
    used in processing the DALI pipeline. The rewriting applies to all functions defined within or called
    from a pipeline definition function. This, among other effects, makes these functions non-serializable.

    Parallel :meth:`external source <nvidia.dali.fn.external_source>` (``parallel=True``), requires that their
    ``source`` argument is serializable. To prevent rewriting of the ``source``, the functions used to create it
    should be decorated with :meth:`@do_not_convert <nvidia.dali.pipeline.do_not_convert>`.
```

Reply:
done